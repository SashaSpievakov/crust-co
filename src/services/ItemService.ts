import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

const itemAPI = createApi({
  reducerPath: 'itemAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Item'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (id) => ({
        url: `/pizzas/${id}`,
      }),
      providesTags: () => ['Item'],
    }),
  }),
});

export default itemAPI;
