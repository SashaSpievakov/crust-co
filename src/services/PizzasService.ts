import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

interface IFetchItems {
  activeCategory: number;
  sortedPropertyName: string;
}

const pizzasAPI = createApi({
  reducerPath: 'pizzasAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: 'https://6344adb1dcae733e8fe3067a.mockapi.io/',
  }),
  tagTypes: ['Pizzas'],
  endpoints: (build) => ({
    fetchPizzas: build.query<IPizzaItem[], IFetchItems>({
      query: ({ activeCategory, sortedPropertyName }) => ({
        url: `pizza-items?${
          activeCategory > 0 ? `category=${activeCategory}&` : ''
        }sortBy=${sortedPropertyName}`,
      }),
      providesTags: () => ['Pizzas'],
    }),
  }),
});

export default pizzasAPI;
