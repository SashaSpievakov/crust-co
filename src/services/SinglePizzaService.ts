import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

export const pizzaAPI = createApi({
  reducerPath: 'pizzaPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Pizza'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (name) => ({
        url: `/pizzas/${name}`,
      }),
      providesTags: (result) =>
        result ? [{ type: 'Pizza', id: result.name }] : ['Pizza'],
    }),
  }),
});
