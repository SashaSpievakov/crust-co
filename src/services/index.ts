export { pizzasAPI } from './PizzasService';
export { pizzaAPI } from './SinglePizzaService';
