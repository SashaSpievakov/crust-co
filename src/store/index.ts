export * from './slices';
export type { AppDispatch, AppStore, RootState } from './store';
export { setupStore } from './store';
