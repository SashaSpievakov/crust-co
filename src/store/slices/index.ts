export type { CartSliceState } from './cart';
export { addItem, clearCart, removeItem, removeItems } from './cart';
export { selectIsLight, setTheme } from './theme';
