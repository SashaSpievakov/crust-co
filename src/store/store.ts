import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import categorySlice from './slices/category/reducer/categoryReducer';
import sortSlice from './slices/sort/reducer/sortReducer';
import searchSlice from './slices/search/reducer/searchValueReducer';
import theme from './slices/theme/reducer/themeReducer';
import cart from './slices/cart/reducer/cartReducer';
import itemAPI from '../services/ItemService';
import pizzasAPI from '../services/PizzasService';

const rootReducer = combineReducers({
  activeCategory: categorySlice,
  activeSort: sortSlice,
  searchValue: searchSlice,
  theme,
  cart,
  [itemAPI.reducerPath]: itemAPI.reducer,
  [pizzasAPI.reducerPath]: pizzasAPI.reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(itemAPI.middleware)
        .concat(pizzasAPI.middleware),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
