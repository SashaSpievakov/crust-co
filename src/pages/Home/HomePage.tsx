import { ChangeEvent, FC, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { ErrorRequest } from 'src/components';
import { useDebounce } from 'src/hooks';
import { Container } from 'src/styles/Base.styled';

import { DropdownSelect, SearchItems } from '../../components/UI';
import { IPizzaItem } from '../../models';
import { pizzasAPI } from '../../services';
import { modifySearchParamsName } from '../../utils/modifySearchParamsName';
import { Categories, ProductsContainer } from './components';
import { Block, Title, Top } from './HomePage.styled';

export const sortNamesArr: string[] = ['rating', 'price', 'A to Z'];

export const HomePage: FC = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const [pizzas, setPizzas] = useState<IPizzaItem[]>([]);
  const [searchValue, setSearchValue] = useState<string>('');
  const debouncedSearchValue = useDebounce(searchValue, 100);
  const [activeSort, setActiveSort] = useState<number>(
    parseInt(searchParams.get('sortBy') || '0', 10) || 0,
  );
  const [activeCategory, setActiveCategory] = useState<number>(
    parseInt(searchParams.get('category') || '0', 10) || 0,
  );

  const sortSearchParam = modifySearchParamsName(sortNamesArr[activeSort]);
  const { data, isSuccess, isError, isLoading } = pizzasAPI.useFetchPizzasQuery(
    {
      activeCategory,
      sortSearchParam,
    },
  );

  useEffect(() => {
    document.title = 'Crust & Co.';
  }, []);

  useEffect(() => {
    if (isSuccess) {
      setPizzas(data);
    }
  }, [activeCategory, sortSearchParam, data, isSuccess]);

  const handleSortSelect = (index: number): void => {
    setActiveSort(index);
    setSearchParams({
      sortBy: index.toString(),
      category: activeCategory.toString(),
    });
  };

  const handleCategorySelect = (index: number): void => {
    setActiveCategory(index);
    setSearchParams({
      category: index.toString(),
      sortBy: activeSort.toString(),
    });
  };

  const onChangeSearch = (e: ChangeEvent<HTMLInputElement>): void => {
    setSearchValue(e.target.value);
  };

  const onClickClose = () => {
    setSearchValue('');
  };

  return isError ? (
    <ErrorRequest />
  ) : (
    <Container data-testid="homePage">
      <Top>
        <Categories
          activeCategory={activeCategory}
          onSelect={handleCategorySelect}
        />
        <DropdownSelect
          chosenValue={sortNamesArr[activeSort]}
          sortNamesArr={sortNamesArr}
          onSelect={handleSortSelect}
        />
      </Top>
      <Block>
        <Title>All Pizzas</Title>
        <SearchItems
          value={searchValue}
          width="240px"
          onChange={onChangeSearch}
          onClickClose={onClickClose}
          expandedWidth="300px"
        />
      </Block>
      <ProductsContainer
        isLoading={isLoading}
        items={pizzas}
        searchValue={debouncedSearchValue}
      />
    </Container>
  );
};
