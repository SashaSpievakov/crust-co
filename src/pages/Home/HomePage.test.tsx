import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import { rest } from 'msw';

import { pizzasAPI } from '../../services';
import { setupStore } from '../../store';
import { renderWithProvidersAndRoutes } from '../../tests/helpers';
import { testServer } from '../../tests/mocks';
import { HomePage } from './HomePage';

describe('HomePage tests', () => {
  beforeAll(() => {
    testServer.listen();
  });

  afterAll(() => {
    testServer.close();
  });

  afterEach(() => {
    testServer.resetHandlers();
    setupStore().dispatch(pizzasAPI.util.resetApiState());
  });

  test('should render the Home page', async () => {
    renderWithProvidersAndRoutes(<HomePage />);
    const lastItem = await screen.findByText(/diablo/i); // eslint-disable-line @typescript-eslint/no-unused-vars

    const loadedHome = screen.getByTestId('homePage');
    expect(loadedHome).toMatchSnapshot();
  });

  test('should render the Home page with a server request error', async () => {
    testServer.use(
      rest.get('*', (_req, res, ctx) => res.once(ctx.status(500))),
    );
    renderWithProvidersAndRoutes(<HomePage />);

    expect(
      await screen.findByText(/something went wrong/i),
    ).toBeInTheDocument();
  });
});
