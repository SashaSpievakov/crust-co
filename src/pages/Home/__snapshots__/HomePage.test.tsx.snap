// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Tests renders the Home page 1`] = `
<article
  class="sc-hhOBVt cSuYdA"
  data-testid="homePage"
>
  <section
    class="sc-fHSyak ekfLmS"
  >
    <section>
      <ul
        class="sc-jIILKH iZcOtu"
      >
        <li
          aria-current="true"
          class="sc-ZqFbI jHNOUM"
          tabindex="-1"
        >
          All
        </li>
        <li
          aria-current="false"
          class="sc-ZqFbI isYAHE"
          tabindex="0"
        >
          Meat
        </li>
        <li
          aria-current="false"
          class="sc-ZqFbI isYAHE"
          tabindex="0"
        >
          Vegetarian
        </li>
        <li
          aria-current="false"
          class="sc-ZqFbI isYAHE"
          tabindex="0"
        >
          Grill
        </li>
        <li
          aria-current="false"
          class="sc-ZqFbI isYAHE"
          tabindex="0"
        >
          Spicy
        </li>
      </ul>
    </section>
    <section
      class="sc-ftTHYK gjTCxi"
    >
      <div
        class="sc-pyfCe kBoKrx"
        tabindex="0"
      >
        <svg
          class="sc-kDvujY iBKcZM"
          data-testid="popupIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.247 11.14 2.451 5.658C1.885 5.013 2.345 4 3.204 4h9.592a1 1 0 0 1 .753 1.659l-4.796 5.48a1 1 0 0 1-1.506 0z"
          />
        </svg>
        <b>
          Sort by
        </b>
        <span
          data-testid="currenChosenSortName"
        >
          rating
        </span>
      </div>
      <section
        aria-hidden="true"
        class="sc-ipEyDJ GIGYJ"
      >
        <ul
          aria-hidden="true"
          data-testid="popupList"
        >
          <li
            class="sc-csuSiG jOMJNz"
            tabindex="0"
          >
            rating
          </li>
          <li
            class="sc-csuSiG iWIHQu"
            tabindex="0"
          >
            price
          </li>
          <li
            class="sc-csuSiG iWIHQu"
            tabindex="0"
          >
            A to Z
          </li>
        </ul>
      </section>
    </section>
  </section>
  <section
    class="sc-fHSyak sc-dubCtV ekfLmS czsuWh"
  >
    <h2
      class="sc-evzXkX cngOxp"
    >
      All Pizzas
    </h2>
    <div
      class="sc-bjfHbI iNTtOj"
    >
      <svg
        class="sc-idXgbr hWNjFA"
        data-testid="searchIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
        />
      </svg>
      <input
        class="sc-dIfARi dxuNau"
        placeholder="Search..."
        value=""
        width="240px"
      />
    </div>
  </section>
  <section
    class="sc-brePNt fdCBqo"
  >
    <article
      class="sc-GKYbw fmZoYD"
    >
      <a
        class="sc-gsGlKL iBkFUq"
        href="/pizzas/Shrimp Pizza"
      >
        <picture>
          <source
            srcset="./assets/img/pizzas/pizza3.webp"
            type="image/webp"
          />
          <source
            srcset="./assets/img/pizzas/pizza3.png"
            type="image/png"
          />
          <img
            alt="Shrimp Pizza pizza"
            class="sc-bWOGAC kVJyAy"
            height="250"
            src="./assets/img/pizzas/pizza3.png"
            width="250"
          />
        </picture>
        <h3
          class="sc-eGugkK fiVuXq"
        >
          Shrimp Pizza
        </h3>
      </a>
      <div
        class="sc-hOzowv iYJRyu"
      >
        <ul>
          <li
            aria-current="true"
            class="sc-dsHJmm ewokfL"
            tabindex="0"
          >
            traditional
          </li>
          <li
            aria-current="false"
            class="sc-dsHJmm dgbJWc"
            tabindex="0"
          >
            thin
          </li>
        </ul>
        <ul>
          <li
            aria-current="true"
            class="sc-dsHJmm ewokfL"
            tabindex="0"
          >
            12
             inch
          </li>
          <li
            aria-current="false"
            class="sc-dsHJmm dgbJWc"
            tabindex="0"
          >
            14
             inch
          </li>
          <li
            aria-current="false"
            class="sc-dsHJmm dgbJWc"
            tabindex="0"
          >
            16
             inch
          </li>
        </ul>
      </div>
      <section
        class="sc-csDkEv iZAaHU"
      >
        <h3
          class="sc-dEVLtI iStDFD"
        >
          10
          $
        </h3>
        <button
          class="sc-bcXHqe sc-eDvSVe sc-gKPRtg glVyWT fwWQTF gmBSoV"
        >
          <svg
            class="sc-kImNAt ifacgh"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
            />
          </svg>
          <span>
            Add
          </span>
        </button>
      </section>
    </article>
    <article
      class="sc-GKYbw fmZoYD"
    >
      <a
        class="sc-gsGlKL iBkFUq"
        href="/pizzas/Pepperoni Pizza"
      >
        <picture>
          <source
            srcset="./assets/img/pizzas/pizza2.webp"
            type="image/webp"
          />
          <source
            srcset="./assets/img/pizzas/pizza2.png"
            type="image/png"
          />
          <img
            alt="Pepperoni Pizza pizza"
            class="sc-bWOGAC kVJyAy"
            height="250"
            src="./assets/img/pizzas/pizza2.png"
            width="250"
          />
        </picture>
        <h3
          class="sc-eGugkK fiVuXq"
        >
          Pepperoni Pizza
        </h3>
      </a>
      <div
        class="sc-hOzowv iYJRyu"
      >
        <ul>
          <li
            aria-current="true"
            class="sc-dsHJmm ewokfL"
            tabindex="0"
          >
            traditional
          </li>
        </ul>
        <ul>
          <li
            aria-current="true"
            class="sc-dsHJmm ewokfL"
            tabindex="0"
          >
            12
             inch
          </li>
          <li
            aria-current="false"
            class="sc-dsHJmm dgbJWc"
            tabindex="0"
          >
            14
             inch
          </li>
        </ul>
      </div>
      <section
        class="sc-csDkEv iZAaHU"
      >
        <h3
          class="sc-dEVLtI iStDFD"
        >
          12
          $
        </h3>
        <button
          class="sc-bcXHqe sc-eDvSVe sc-gKPRtg glVyWT fwWQTF gmBSoV"
        >
          <svg
            class="sc-kImNAt ifacgh"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
            />
          </svg>
          <span>
            Add
          </span>
        </button>
      </section>
    </article>
    <article
      class="sc-GKYbw fmZoYD"
    >
      <a
        class="sc-gsGlKL iBkFUq"
        href="/pizzas/Diablo"
      >
        <picture>
          <source
            srcset="./assets/img/pizzas/pizza6.webp"
            type="image/webp"
          />
          <source
            srcset="./assets/img/pizzas/pizza6.png"
            type="image/png"
          />
          <img
            alt="Diablo pizza"
            class="sc-bWOGAC kVJyAy"
            height="250"
            src="./assets/img/pizzas/pizza6.png"
            width="250"
          />
        </picture>
        <h3
          class="sc-eGugkK fiVuXq"
        >
          Diablo
        </h3>
      </a>
      <div
        class="sc-hOzowv iYJRyu"
      >
        <ul>
          <li
            aria-current="true"
            class="sc-dsHJmm ewokfL"
            tabindex="0"
          >
            traditional
          </li>
          <li
            aria-current="false"
            class="sc-dsHJmm dgbJWc"
            tabindex="0"
          >
            thin
          </li>
        </ul>
        <ul>
          <li
            aria-current="true"
            class="sc-dsHJmm ewokfL"
            tabindex="0"
          >
            12
             inch
          </li>
          <li
            aria-current="false"
            class="sc-dsHJmm dgbJWc"
            tabindex="0"
          >
            16
             inch
          </li>
        </ul>
      </div>
      <section
        class="sc-csDkEv iZAaHU"
      >
        <h3
          class="sc-dEVLtI iStDFD"
        >
          15
          $
        </h3>
        <button
          class="sc-bcXHqe sc-eDvSVe sc-gKPRtg glVyWT fwWQTF gmBSoV"
        >
          <svg
            class="sc-kImNAt ifacgh"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
            />
          </svg>
          <span>
            Add
          </span>
        </button>
      </section>
    </article>
  </section>
</article>
`;
